### PR TITLE
doc: Add BSD library top level doc

### PR DESCRIPTION
**Pull request by b-gent**
First version of the RST file.
Adding svg diagrams for 'overview' and for 'architecture'.
This version of the RST does not yet include a 'Security tag model'
section which will be added soon.

Signed-off-by: Bartosz Gentkowski <bartosz.gentkowski@nordicsemi.no>

### CODE REVIEW

**Comment by b-gent** [comment]
Attaching output for your convenience
[BSD Library — nrfxlib 1.0.99 documentation.pdf](https://github.com/NordicPlayground/nrfxlib/files/3851748/BSD.Library.nrfxlib.1.0.99.documentation.pdf)

File: bsdlib/README.rst [hunk 1]
```diff
@@ -3,9 +3,131 @@
...
+* AT commands
+* GNSS
+* AGPS
+* PDN management
+* Modem DFU
+
+.. note::
+   The DNS address management protocol is not yet implemented.
+
+The library implements the following standard BSD functions:
```
Comment (lemrey) [inline bsdlib/README.rst]: While this is a complete list, each socket supports its own subset of this API e.g. the modem DFU socket doesn't support `listen` or `poll` and it doesn't make sense for it to support those either. Instead of mentioning that `select` is not yet implemented (line 67), we could either provide a more exhaustive list of which functions are supported/implemented on each socket, or just have a note saying that each socket supports a subset of this interface.

File: bsdlib/README.rst [hunk 2]
```diff
@@ -3,9 +3,131 @@
...
+
+The BSD library has been designed in such way that you can port it to any OS.
+Therefore, procedures that would normally require some kind of OS interaction, like going into sleep and IRQ reprioritization, have been moved out of the library to an open :file:`c` file in which OS-specific integration  can be implemented.
+For more information, see :ref:`bsdlib_ug_porting`.
+
+Limitations
+***********
+
+GNSS, AGPS, and Modem DFU are currently not part of the generic sockets.
+Therefore, operating on them using ``poll`` is not possible yet.
```
Comment (lemrey) [inline bsdlib/README.rst]: There are also several other function calls that are not supported; see my comment above.

File: bsdlib/README.rst [hunk 3]
```diff
@@ -3,9 +3,131 @@
...
+This might change in future releases of the BSD library.
```
Comment (lemrey) [inline bsdlib/README.rst]: Or it might not.. this line could probably be removed.

File: bsdlib/README.rst [hunk 4]
```diff
@@ -3,9 +3,131 @@
...
+
+PDN sockets only support ``open``, ``close``, ``setsockopt``, and ``getsockopt`` in order to configure a PDN with a APN name.
```
Comment (lemrey) [inline bsdlib/README.rst]: This is a complete list; it can stay.

File: bsdlib/README.rst [hunk 5]
```diff
@@ -3,9 +3,131 @@
 BSD Library
 ###########
 
+The BSD library is Nordic Semiconductor's implementation of the BSD Socket API, which is a set of standard function calls that can be used in an application.
```
Comment (ru-fu) [inline bsdlib/README.rst]: Can we add a link to BSD Socket API?
(For example, https://en.wikipedia.org/wiki/Berkeley_sockets#Socket_API_functions )

File: bsdlib/README.rst [hunk 6]
```diff
@@ -3,9 +3,131 @@
 BSD Library
```
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
BSD library
```

File: bsdlib/README.rst [hunk 7]
```diff
@@ -3,9 +3,131 @@
...
 ###########
 
+The BSD library is Nordic Semiconductor's implementation of the BSD Socket API, which is a set of standard function calls that can be used in an application.
+This library aims to be compatible with the BSD socket standard to simplify writing and porting of existing applications. 
+
+The library has the following set of sockets:
+
+* 8 generic sockets that can be used generically between UDP, TCP, TLS, DTLS, and AT commands.
```
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
* 8 generic sockets that can be used generically between UDP, TCP, TLS, DTLS, and AT commands
```

File: bsdlib/README.rst [hunk 8]
```diff
@@ -3,9 +3,131 @@
...
+* 1 GNSS/AGPS socket
+* 1 Modem DFU socket
```
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
* 1 modem DFU socket
```

File: bsdlib/README.rst [hunk 9]
```diff
@@ -3,9 +3,131 @@
...
+Overview
+********
+
+.. figure:: doc/images/bsd_lib_overview.svg
+   :alt: BSD library overview diagram
+
+   BSD library overview diagram
+
+The BSD library is the primary interface to operate the nRF9160 modem in order to establish LTE-M, NBIOT, and GNSS connections.
+The library abstracts away the details on memory management and synchronization leaving the application to focus on writing the application using standardized API calls.
```
Comment (ru-fu) [inline bsdlib/README.rst]: leaving the application to focus on writing the application?

File: bsdlib/README.rst [hunk 10]
```diff
@@ -3,9 +3,131 @@
...
+
+.. note::
+   The ``select`` function is not yet implemented.
+
+
+Library internals
+*****************
+
+The purpose of the BSD library is to communicate with the nRF9160 modem firmware.
+This is done by sending events and getting responses over the Inter Processor Communication peripheral (IPC).
```
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
This is done by sending events and getting responses over the Inter Processor Communication (IPC) peripheral.
```

File: bsdlib/README.rst [hunk 11]
```diff
@@ -3,9 +3,131 @@
...
+To facilitate data transfer between the application and the modem, payload data is set in the shared RAM.
+
+The modem contains a full IP and DTLS/TLS stack as well as GNSS.
```
Comment (ru-fu) [inline bsdlib/README.rst]: Can we add links and/or spelled out versions? I have no idea what GNSS is.

File: bsdlib/README.rst [hunk 12]
```diff
@@ -3,9 +3,131 @@
...
+The library internally communicates over a Remote Procedure Call (RPC) protocol using the IPC and the shared RAM to utilize the modules inside the modem.
+The library provides address families, types, and protocols that dispatch the protocols to the right backend in the modem and thus communicates with this backend.
```
Comment (ru-fu) [inline bsdlib/README.rst]: Are there different backends in the modem? Is this relevant to the user?

File: bsdlib/README.rst [hunk 13]
```diff
@@ -3,9 +3,131 @@
...
+
+.. figure:: doc/images/bsd_lib_architecture.svg
+   :alt: BSD library architecture diagram
+
+   BSD library architecture diagram
+
+Porting flexibility
+*******************
+
+The BSD library has been designed in such way that you can port it to any OS.
```
Comment (ru-fu) [inline bsdlib/README.rst]: What do we mean with OS here? Zephyr, FreeRTOS, and so on?

File: bsdlib/README.rst (hunk 4 above)
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
PDN sockets only support ``open``, ``close``, ``setsockopt``, and ``getsockopt`` to configure a PDN with an APN name.
```

File: bsdlib/README.rst [hunk 14]
```diff
@@ -3,9 +3,131 @@
...
+* TCP max packet size of 4096
+* TLS internal buffer size of 2.5 KB
+* Only one DTLS/TLS handshake can be done at the time.
+  No parallel handshake is possible.
+  
+Modem limitations on the built-in key storage:
+
+* Max number of assets is 16 (slots).
+* 1 PSK takes 1 slot.
+* 1 Identity takes 1 slot.
```
Comment (ru-fu) [inline bsdlib/README.rst]:
```suggestion
* 1 identity takes 1 slot.
```

File: bsdlib/README.rst [hunk 15]
```diff
@@ -3,9 +3,131 @@
...
+* 1 certificate might span over multiple slots depending on the size.
+  Testing is needed to verify that the certificate of choice gets into the modem and can be used by referencing the security tag.
```
Comment (ru-fu) [inline bsdlib/README.rst]: gets stored in the modem?

File: bsdlib/README.rst (hunk 5 above)
Comment (ru-fu) [reply bsdlib/README.rst]: Oh, and do we know what BSD actually stands for? In the picture it says Berkeley Socket Descriptors, but Google has never heard of this term ...

File: bsdlib/README.rst (hunk 1 above)
Comment (b-gent) [reply bsdlib/README.rst]: Changing the note at the end of the list to:
"Each socket supports a subset of these functions."

File: bsdlib/README.rst (hunk 3 above)
Comment (b-gent) [reply bsdlib/README.rst]: yes

File: bsdlib/README.rst [hunk 16]
```diff
@@ -3,9 +3,131 @@
...
+
+.. figure:: doc/images/bsd_lib_overview.svg
+   :alt: BSD library overview diagram
+
+   BSD library overview diagram
+
+The BSD library is the primary interface to operate the nRF9160 modem in order to establish LTE-M, NBIOT, and GNSS connections.
+The library abstracts away the details on memory management and synchronization leaving the application to focus on writing the application using standardized API calls.
+
+The current implementation of the BSD library supports the following protocols:
```
[the remark this answers is not in the document]
Comment (b-gent) [reply bsdlib/README.rst]: Changed to 'features and protocols'.
Security tag model (name might change) will be a separate chapter on this page presenting some details, but it is postponed to after 1.1.0.

File: bsdlib/README.rst (hunk 5 above)
Comment (evenl) [reply bsdlib/README.rst]: BSD seems to stand for just Berkely Sockets in this case, or that seems be what everyone refer to, so I am not sure where the D is lost.

File: bsdlib/README.rst (hunk 7 above)
Comment (evenl) [reply bsdlib/README.rst]: I think maybe "freely" would have been a better word than generically.

File: bsdlib/README.rst (hunk 11 above)
Comment (b-gent) [reply bsdlib/README.rst]: postponed

File: bsdlib/README.rst (hunk 12 above)
Comment (b-gent) [reply bsdlib/README.rst]: postponed

File: bsdlib/README.rst (hunk 13 above)
Comment (b-gent) [reply bsdlib/README.rst]: postponed, will provide examples (but yes, Zephyr is one of them)

File: bsdlib/README.rst (hunk 15 above)
Comment (b-gent) [reply bsdlib/README.rst]: changed to:
'is correctly stored in the modem'

File: bsdlib/README.rst (hunk 9 above)
Comment (evenl) [reply bsdlib/README.rst]: The library abstracts away memory management details and application synchronization, leaving the developer to focus on writing the application using standardized API calls.

File: bsdlib/README.rst (hunk 5 above)
Comment (evenl) [reply bsdlib/README.rst]: Actually BSD stands for Berkeley Software Distribution, so it's BSD Sockets is a part of Berkeley Software Distribution interfaces.

File: bsdlib/README.rst (hunk 9 above)
Comment (b-gent) [reply bsdlib/README.rst]: Changed to:
'The library abstracts away the details on memory management and synchronization allowing you to focus on writing the application using standardized API calls.'

File: bsdlib/README.rst (hunk 12 above)
Comment (evenl) [reply bsdlib/README.rst]: I am not sure if this sentence make sense at all:

Maybe something like this:

By providing the right combinations of address families, types, and protocols to the library, will the library be able to dispatch data  to and from the different backends in the modem firmware.


that dispatch the protocols to the right backend in the modem and thus communicates with this backend.

File: bsdlib/README.rst (hunk 5 above)
Comment (b-gent) [reply bsdlib/README.rst]: That's interesting. I'll just change to BSD in the picture to be on the safe side and follow up with Glenn later.

File: bsdlib/README.rst (hunk 13 above)
Comment (evenl) [reply bsdlib/README.rst]: Yes, it's OS agnostic. So it's meant to work with any OS with little effort.

File: bsdlib/README.rst (hunk 12 above)
Comment (b-gent) [reply bsdlib/README.rst]: Changed to:
"You can dispatch data to and from the different modules in the modem firmware by providing the right combinations of address families, types, and protocols to the BSD library."

@evenl  please let me know if this is ok

File: bsdlib/README.rst (hunk 7 above)
Comment (b-gent) [reply bsdlib/README.rst]: @evenl sounds good, thanks

File: bsdlib/README.rst [hunk 17]
```diff
@@ -1,11 +1,134 @@
...
+
+The purpose of the BSD library is to communicate with the nRF9160 modem firmware.
+This is done by sending events and getting responses over the Inter Processor Communication (IPC) peripheral.
+To facilitate data transfer between the application and the modem, payload data is set in the shared RAM.
+
+The modem contains a full IP and DTLS/TLS stack as well as GNSS.
+The library internally communicates over a Remote Procedure Call (RPC) protocol using the IPC and the shared RAM to utilize the modules inside the modem.
+You can dispatch data to and from the different modules in the modem firmware by providing the right combinations of address families, types, and protocols to the BSD library.
+
+The library provides address families, types, and protocols that dispatch the protocols to the right backend in the modem and thus communicates with this backend.
```
Comment (ru-fu) [inline bsdlib/README.rst]: Now the sentence is in here twice. ;)

File: bsdlib/README.rst [hunk 18]
```diff
@@ -1,11 +1,134 @@
...
+
+The BSD library aims to provide a standard programming interface on top of all of these backends so that the application developer does not have to manage all intricate details of the low-level parameters passed back and forth between the application and the modem.
```
Comment (ru-fu) [inline bsdlib/README.rst]: Here we're referring back to the backends - change to "modules" as well?

File: bsdlib/README.rst (hunk 17 above)
Comment (b-gent) [reply bsdlib/README.rst]: yes, corrected

File: bsdlib/README.rst (hunk 18 above)
Comment (b-gent) [reply bsdlib/README.rst]: yes

**Comment by b-gent** [comment]
Most of the comments are addressed, the ones that are left are postponed.
Branch is rebased.
Unless there is something critical left to fix, I need someone to give this a green tick and merge.
@rlubos  @ru-fu  @lemrey

**Comment by rlubos** [comment]
I should've asked earlier, please open a manifest update PR in -nrf for this.
